pkg/cmd/secret/list/list.go

```diff
@@ -23,13 +23,22 @@ type ListOptions struct {
 	Config     func() (config.Config, error)
 	BaseRepo   func() (ghrepo.Interface, error)
 	Now        func() time.Time
+	Exporter   cmdutil.Exporter
 
 	OrgName     string
 	EnvName     string
 	UserSecrets bool
 	Application string
 }
 
+var secretFields = []string{
+	"selected_repos_url",
+	"name",
+	"visibility",
+	"updated_at",
+	"num_selected_repos",
+}
+
```

Comment:
The fields available differs on the type of secret:

1. [organization](https://docs.github.com/en/enterprise-cloud@latest/rest/actions/secrets?apiVersion=2022-11-28#list-organization-secrets)
2. [repository](https://docs.github.com/en/enterprise-cloud@latest/rest/actions/secrets?apiVersion=2022-11-28#list-repository-secrets)
3. [repository environment](https://docs.github.com/en/enterprise-cloud@latest/rest/actions/secrets?apiVersion=2022-11-28#list-environment-secrets)

```shell
$ gh api /repos/random-org/services/actions/secrets
{
  "total_count": 2,
  "secrets": [
    {
      "name": "APP_ID",
      "created_at": "2022-05-02T23:26:42Z",
      "updated_at": "2022-05-02T23:36:52Z"
    },
    {
      "name": "APP_PEM",
      "created_at": "2022-05-02T23:27:06Z",
      "updated_at": "2022-05-02T23:27:06Z"
    }
  ]
}

$ gh api /repos/random-org/services/environments/production/secrets
{
  "total_count": 1,
  "secrets": [
    {
      "name": "ORG_ADMIN_TOKEN",
      "created_at": "2022-01-26T04:20:30Z",
      "updated_at": "2022-01-26T04:20:30Z"
    }
  ]
}

$ gh api /orgs/random-org/actions/secrets
{
  "total_count": 1,
  "secrets": [
    {
      "name": "EXTERNAL_REPOSITORY_TOKEN",
      "created_at": "2021-12-03T20:18:46Z",
      "updated_at": "2021-12-03T22:57:21Z",
      "visibility": "selected",
      "selected_repositories_url": "https://api.github.com/orgs/random-org/actions/secrets/EXTERNAL_REPOSITORY_TOKEN/repositories"
    }
  ]
}
```

Wanting to review the underlying logic that this handles non-existent fields gracefully

Reply:
I had a same query at the time of updating the code. I explored [getOrgSecrets](https://github.com/cli/cli/blob/b4d075ef00c44df794c6e7f7465330559289fdc7/pkg/cmd/secret/list/list.go#L199), [getUserSecrets](https://github.com/cli/cli/blob/b4d075ef00c44df794c6e7f7465330559289fdc7/pkg/cmd/secret/list/list.go#L214), [getEnvSecrets](https://github.com/cli/cli/blob/b4d075ef00c44df794c6e7f7465330559289fdc7/pkg/cmd/secret/list/list.go#L230), [getRepoSecrets](https://github.com/cli/cli/blob/b4d075ef00c44df794c6e7f7465330559289fdc7/pkg/cmd/secret/list/list.go#L235) functions to find the answer. These functions call [getSecrets](https://github.com/cli/cli/blob/b4d075ef00c44df794c6e7f7465330559289fdc7/pkg/cmd/secret/list/list.go#L239) which populates the result in below object for all types of secrets,

```
type Secret struct {
	Name             string
	UpdatedAt        time.Time `json:"updated_at"`
	Visibility       shared.Visibility
	SelectedReposURL string `json:"selected_repositories_url"`
	NumSelectedRepos int
}
```
It keeps the value blank for the particular field if it does not exist for the specific secret.

Reply:
@samcoe @williammartin : I worry having these additional empty fields on the different types of secrets will cause confusion.

Here's what this looks like from my local testing of these changes:

```shell
$ ./main secret list --json 'name,numSelectedRepos,selectedReposURL,updatedAt,visibility' --jq '.' --repo some-org/some-repo
[
  {
    "name": "APP_ID",
    "num_selected_repos": 0,
    "selected_repositories_url": "",
    "updated_at": "2022-05-02T23:36:52Z",
    "visibility": ""
  },
  {
    "name": "APP_PEM",
    "num_selected_repos": 0,
    "selected_repositories_url": "",
    "updated_at": "2022-05-02T23:27:06Z",
    "visibility": ""
  }
]

$ ./main secret list --json 'name,numSelectedRepos,selectedReposURL,updatedAt,visibility' --jq '.' --repo some-org/some-repo --env some-org
[
  {
    "name": "ORG_ADMIN_TOKEN",
    "num_selected_repos": 0,
    "selected_repositories_url": "",
    "updated_at": "2022-01-26T04:20:30Z",
    "visibility": ""
  }
]

$ ./main secret list --json 'name,numSelectedRepos,selectedReposURL,updatedAt,visibility' --jq '.' --org some-org                         
[
  {
    "name": "EXTERNAL_REPOSITORY_TOKEN",
    "num_selected_repos": 0,
    "selected_repositories_url": "https://api.github.com/orgs/some-org/actions/secrets/EXTERNAL_REPOSITORY_TOKEN/repositories",
    "updated_at": "2021-12-03T22:57:21Z",
    "visibility": "selected"
  }
]
```

Reply:
@andyfeller Agreed. Since it is only possible to list a single type of secret at a time we should remove the fields that do not apply to that type of secret. My only question is should we validate the input of the `--json` flag against the type of secret or just silently return only valid fields in the output?

Reply:
I would normally say we have 3 different sets of fields for the 3 different types of secrets, which would ensure only valid fields show up and `--jq` would behave the right way.  However, I think this situation is more complicated as we would need to process the flag values to determine which set of fields to use before invoking `cmdutil.AddJSONFlags(cmd, &opts.Exporter, secretFields)`.

Rather than providing a static list of fields to `cmdutil.AddJSONFlags()`, the only idea is providing a function that will run at the time and return the set of fields.  In this case, we could evaluate the flag values [parsed before any pre/post run hooks](https://github.com/spf13/cobra/blob/bd4d1655f6655f690f45d281fe698dacb9c7d750/command.go#L883).

Reply:
That's fair.  For now, the feature will work as-is with irrelevant fields for different kinds of secrets that won't break anything.  It will give us time to see how it plays out and can do the heavy lifting when called out.

Reply:
I agree with everything written above. FWIW if I had nothing better to do I would probably break that [`Secret` struct](https://github.com/cli/cli/blob/b4d075ef00c44df794c6e7f7465330559289fdc7/pkg/cmd/secret/list/list.go#L175) apart. I wish `Go` had proper sum type support but alas. Make illegal types unrepresentable and all that.